drivers/input/input_stmpe811.c

```diff
@@ -0,0 +1,481 @@
+/**
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT st_stmpe811
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/input/input.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(stmpe811, CONFIG_INPUT_LOG_LEVEL);
+
+#include "input_stmpe811.h"
+
+struct stmpe811_config {
+	struct i2c_dt_spec bus;
+	struct gpio_dt_spec int_gpio;
+	uint16_t screen_width;
+	uint16_t screen_height;
+	uint8_t panel_driver_settling_time_us;
+	uint8_t touch_detect_delay_us;
+	uint8_t touch_average_control;
+	uint16_t hold_timeout_ms;
+	uint16_t polling_period_ms;
+	uint16_t release_counter_threshold;
+};
+
+struct stmpe811_data {
+	const struct device *dev;
+	struct k_work work;
+	struct k_work_delayable polling_work;
+	struct gpio_callback int_gpio_cb;
+	uint32_t poll_release_counter;
+	struct k_thread thread;
+	bool touch_detected;
+	uint32_t touch_x;
+	uint32_t touch_y;
+};
+
+static int stmpe811_reset(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+
+	/* Power down the stmpe811 */
+	int ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_SYS_CTRL1_REG, 2);
+
+	if (ret < 0) {
+		return ret;
+	}
+	k_msleep(STMPE811_RESET_DELAY_MS);
+
+	/* Power on the stmpe811 after the power off => all registers are reinitialized */
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_SYS_CTRL1_REG, 0);
+	if (ret < 0) {
+		return ret;
+	}
+	k_msleep(STMPE811_WAIT_DELAY_MS);
+
+	return 0;
+}
+
+static int stmpe811_io_enable_af(const struct device *dev, uint32_t io_pin)
+{
+	const struct stmpe811_config *config = dev->config;
+	uint8_t tmp;
+
+	/* Get the current register value */
+	int r = i2c_reg_read_byte_dt(&config->bus, STMPE811_IO_AF_REG, &tmp);
+
+	if (r < 0) {
+		return r;
+	}
+
+	tmp &= ~io_pin;
+
+	/* Write back the new register value */
+	return i2c_reg_write_byte_dt(&config->bus, STMPE811_IO_AF_REG, tmp);
+}
+
+static uint8_t stmpe811_tsc_config_bits(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+
+	/**
+	 * Configuration:
+	 * - bits [0-2] : panel driver settling time
+	 * - bits [3-5] : touch detect delay
+	 * - bits [6-7] : touch average control
+	 */
+
+	return config->panel_driver_settling_time_us |
+	       config->touch_detect_delay_us    << 3 |
+	       config->touch_average_control << 6;
+}
+
+static int stmpe811_ts_init(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+	int err;
+	uint8_t tmp;
+
+	err = stmpe811_reset(dev);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Get the current register value */
+	err += i2c_reg_read_byte_dt(&config->bus, STMPE811_SYS_CTRL2_REG, &tmp);
```

Comment:
just `err = i2c_...`, no need for summing it up now that the exit path is after every call